go/ngql/query_test.go

```diff
@@ -237,6 +243,32 @@ func (suite *QueryGraphQLSuite) TestListOfUnionOfStructs() {
 	suite.assertQueryResult(list, "{root{elements{... on FooStruct{a b} ... on BarStruct{b} ... on BazStruct{c}}}}", `{"data":{"root":{"elements":[{"a":28,"b":"baz"},{"b":"bar"},{"c":true}]}}}`)
 }
 
+func (suite *QueryGraphQLSuite) TestListOfUnionOfStructsConflictingFieldTypes() {
+	list := types.NewList(
+		types.NewStruct("Foo", types.StructData{
+			"a": types.Number(28),
+		}),
+		types.NewStruct("Bar", types.StructData{
+			"a": types.String("bar"),
+		}),
+		types.NewStruct("Baz", types.StructData{
```

Comment:
The GraphQL spec says that all fields of a struct are always optional. Thus, I think the thing to do will be to define the struct as the union of all the fields that it might have -- which as chance has it it what the type simplification kind of does. For this usage, we'd have to union not just field instances within a collection, but all usages of that Struct reachable from a given value.